src/SAWScript/Prover/MRSolver/Solver.hs

```diff
@@ -1235,67 +1249,65 @@ mrRefinesFunH k _ _ t1 _ t2 = k t1 t2
 askMRSolverH f t1 t2 =
   do mrUVars >>= mrDebugPPPrefix 1 "askMRSolverH uvars:"
      m1 <- normCompTerm t1
      m2 <- normCompTerm t2
      f m1 m2
-     case (m1, m2) of
-       -- If t1 and t2 are both named functions, our result is the opaque
-       -- FunAssump that forall xs. f1 xs |= f2 xs'
-       (FunBind f1 args1 (CompFunReturn _ _),
-        FunBind f2 args2 (CompFunReturn _ _)) ->
-         mrUVars >>= \uvar_ctx ->
-         return $ Just (f1, FunAssump { fassumpCtx = uvar_ctx,
-                                        fassumpArgs = args1,
-                                        fassumpRHS = OpaqueFunAssump f2 args2 })
-       -- If just t1 is a named function, our result is the rewrite FunAssump
-       -- that forall xs. f1 xs |= m2
-       (FunBind f1 args1 (CompFunReturn _ _), _) ->
-         mrUVars >>= \uvar_ctx ->
-         return $ Just (f1, FunAssump { fassumpCtx = uvar_ctx,
-                                        fassumpArgs = args1,
-                                        fassumpRHS = RewriteFunAssump m2 })
-       _ -> return Nothing
-
--- | Test two monadic, recursive terms for refinement. On success, if the
--- left-hand term is a named function, returning a 'FunAssump' to add to the
--- 'MREnv'.
+
+-- | Test two monadic, recursive terms for refinement
 askMRSolver ::
   SharedContext ->
   MREnv {- ^ The Mr Solver environment -} ->
   Maybe Integer {- ^ Timeout in milliseconds for each SMT call -} ->
+  (Set VarIndex -> Sequent -> TopLevel (SolverStats, SolveResult)) {- ^ ... -} ->
+  Refnset t {- ^ Any additional refinements to be assumed by Mr Solver -} ->
   [(LocalName, Term)] {- ^ Any universally quantified variables in scope -} ->
-  Term -> Term -> IO (Either MRFailure MRSolverResult)
-askMRSolver sc env timeout args t1 t2 =
-  runMRM sc timeout env $
+  Term -> Term -> TopLevel (Either MRFailure (SolverStats, MREvidence t))
+askMRSolver sc env timeout askSMT rs args t1 t2 =
+  execMRM sc timeout env askSMT rs $
   withUVars (mrVarCtxFromOuterToInner args) $ \_ ->
     do tp1 <- liftIO $ scTypeOf sc t1 >>= scWhnf sc
        tp2 <- liftIO $ scTypeOf sc t2 >>= scWhnf sc
        mrDebugPPPrefixSep 1 "mr_solver" t1 "|=" t2
        mrRefinesFunH (askMRSolverH mrRefines) [] tp1 t1 tp2 t2
 
--- | Return the 'FunAssump' to add to the 'MREnv' that would be generated if
--- 'askMRSolver' succeeded on the given terms.
-assumeMRSolver ::
+-- | The continuation passed to 'mrRefinesFunH' in 'refinementTerm' - returns
+-- the 'Term' which is the refinement (@Prelude.refinesS@) of the given
+-- 'Term's, after quantifying over all current 'mrUVars' with Pi types
+refinementTermH :: Term -> Term -> MRM t Term
+refinementTermH t1 t2 =
+  do (SpecMParams ev1 stack1, tp1) <- fromJust . asSpecM <$> mrTypeOf t1
+     (SpecMParams ev2 stack2, tp2) <- fromJust . asSpecM <$> mrTypeOf t2
+     rpre <- liftSC2 scGlobalApply "Prelude.eqPreRel" [ev2, stack2]
+     rpost <- liftSC2 scGlobalApply "Prelude.eqPostRel" [ev2, stack2]
```

Comment:
Just double checking, are these supposed to have the same arguments (`[ev2, stack2]`)? Or is one supposed to have `[ev1, stack1]` as an argument? I'm not familiar enough with the codebase to know, but just wanted to double check that this isn't a copy/paste error.

Reply:
Good catch! I left some notes on this function that explain: in fact one of the line below the ones you highlighted will assert that `ev1 == ev2` and `stack1 == stack2`, so it actually doesn't matter. I hopefully made that clearer by adding some comments and using `ev2` and `stack2` consistently throughout.